### DISCUSSION TITLE
Lockdown inner workings in case of device disconnection

### DISCUSSION THREAD

**Discussion by icyflatline** [doronz88/pymobiledevice3#399] [Q&A] [answered]
Hi,
I'd like to understand some specifics of the lockdown service and LockdownClient functionality.

In case of USB connection, is there a way for a lockdown to identify device disconnection from host? For example, due to physical disconnection with USB cable, or in case of device restart (sending 'restart' command or kernel panic)?

P.S. 
From the implementation of the watch() method of the CrashReportsManager class under pymobiledevice3/services/crash_reports.py, it seems like it can detect creation of the kernel panic crash report. As far as I understand, panic log is written after kernel panic occurrence, meaning - after device restart. So, in this case, does lockdown somehow detects device disconnection and then tries to reestablish connection to host? If so, how does this mechanism work?

Thanks in advance.

**Comment by doronz88** [comment, marked as answer]
When a `ConnectionAbortedError` you can deduce the device was disconnected. However, the better way to do this would be through `usbmux`. You can `list_devices()` in a loop to detect attaches and detaches.

Regarding the panics, after the device reboots `DumpPanic` will create the crash report inside the device's crash reports folder and then pymobiledevice3 can reconnect and pull the crash files

**Reply by icyflatline** [reply]
Thanks for the tip about using the `list_devices()` from `usbmux`.

Regarding the panics.
1. `DumpPanic` - is that the name of the process that collects panic crash report from device's NVRAM and writes it on device storage?
2. How does the device reboot is handled for instance in the aforementioned `watch()` method under `pymobiledevice3/services/crash_reports.py` in case of panic? Does `OsTraceService(lockdown=self.lockdown).syslog()` continue to generate syslog messages, even though device is being rebooted?

**Reply by doronz88** [reply]
1. Exactly
2. You'll get a `ConnectionAbortedError` and you'll have to create a new lockdown connection

**Reply by icyflatline** [reply]
Awesome,
Thank you!
